### ISSUE TITLE
Private Practitioner: E-filing a document in a consolidated case

### ISSUE THREAD

**Issue by cholly75** [ustaxcourt/ef-cms#8535]
As a private practitioner in a case, to streamline the filing process in a consolidated case group, I need to be able to file a single document across the consolidated case group as a single operation.

With the recent enhancements to DAWSON to better handle consolidated groups of cases, we would like to now make available to practitioners the ability to e-file a document across multiple cases in their consolidated group. This feature should be available to both private and IRS practitioners.

## Pre-Conditions
ustaxcourt/ef-cms-flexion#9597

## Acceptance Criteria
* When e-filing an eligible document (see Notes section below re: eligible document types) in a case to which they are a party, if that case is a part of a consolidated group (Lead or Member) a practitioner is presented with the option to file in all cases across the consolidated group to which that case belongs
* The default option for the selection UI is to file in all cases in the consolidated group
* When filing in multiple cases:
  - Document entries on case docket records are created for each case selected by Practitioner
  - Cover sheets are applied to the document on each docket entry in the following manner for consolidated cases:
      - Petitioner name(s) of the *Lead case in the group* appended with "et al"
      - Docket number and document number of each case selected are listed in ascending order by docket number
  - Service is performed on all parties in all cases being filed in
  - QC work items are spawned for each docket entry created by the filing
  - The FiledBy value on the docket record for each entry indicates the party filing the document, even if they are not a party to a case in which the document is being docketed
  - Any attachments or supporting documentation for the document being filed are also filed with that document in all cases where docketed
  - Messaging throughout Practitioner flow is updated to reflect multiple cases are being filed in
  - At the final step in filing document on the Review Page, Practitioner is presented with a reminder of all cases the document will be filed in
  - At the final step in filing document, Practitioner can cancel out of the workflow
  - The document created by clicking the Print Receipt link displayed after completing filing reflects all docket numbers in which the document was filed and appends "et al.," to the Lead Case Petitioner names in the caption.
* The following document types should not be multi-docketed with this functionality (i.e., do not invoke this if the document being filed is one of the following event codes:
ACED
COED
M007
M083
PSDE

## Security Considerations
* Does this work make you nervous about privacy or security?
* Does this work make major changes to the system?
* Does this work implement new authentication or security controls?
* Does this work create new methods of authentication, modify existing security controls, or explicitly implement any security or privacy features?

## Notes
* Should tie to the same feature flag used by ustaxcourt/ef-cms#8252
* Need to test w/ IRS service agent
* Eligible document types include all documents currently available to be e-filed by Practitioners
* Should work on mobile
* Can we show progress in filing/serving each docket entry to the user as opposed to a spinning wheel?
* Solution will need to be re-used/consistent w/ IRS practitioner solution

## Notes
* Should tie to the same feature flag used by ustaxcourt/ef-cms#8186/#9595

## Story Definition of Ready (updated on 12/23/22)
The following criteria must be met in order for the user story to be picked up by the Flexion development team.
The user story must: 
- [ ]  Is framed in business/user need, the value has been addressed.
- [ ] Includes acceptance criteria
- [ ] Has been refined
- [ ] Pre conditions have been satisfied. 
**Process:** 
Flexion developers and designers will test if the story meets acceptance criteria and test cases in Flexion dev and staging environments (?standard testing?). If additional acceptance criteria or testing scenarios are discovered while the story is in progress, a new story should be created, added to the backlog and prioritized by the product owner. 

## Definition of Done (Updated 5-19-22)
**Product Owner**
 - [ ]  Acceptance criteria have been met and validated on the Court's migration environment
 - [ ] Add scenario to testing document, if applicable (https://docs.google.com/spreadsheets/d/1FUHKC_YrT-PosaWD5gRVmsDzI1HS_U-8CyMIb-qX9EA/edit?usp=sharing)

**UX**
 - [ ] Business test scenarios have been refined to meet all acceptance criteria 
 - [ ] Usability has been validated
 - [ ] Wiki has been updated (if applicable) 
 - [ ] Story has been tested on a mobile device (for external users only)

**Engineering**
 - [ ] Automated test scripts have been written, including visual tests for newly added PDFs
 - [ ] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [ ] Verify that language for docket record for internal users and external users is identical
 - [ ] New screens have been added to pa11y scripts
 - [ ] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [ ] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated
 - [ ] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [ ] Interactors should validate entities before calling persistence methods
 - [ ] Code refactored for clarity and to remove any known technical debt
 - [ ] Acceptance criteria for the story has been met
 - [ ] If there are special instructions in order to deploy into the next environment, add them as a comment in the story
 - [ ] Reviewed by UX on a deployed environment.
 - [ ] Reviewed by PO on a deployed environment. Can be deployed to the Court's `test` environment if prod-like data is required. Otherwise deployed to any `experimental` environment.
 - [ ] Deployed to the Court's `staging` environment.

**Comment by TomElliottFlexion** [comment]
Note: In order to allow privatePractitioners to request access to consolidated groups of cases, this story gives their version of the PublicCase entity access to the the petitioners, irsPractitioners, and privatePractitioners arrays

Edit: This has been reverted as privatePractitioners no longer have the ability to request access to consolidated groups

**Comment by ttlenard** [comment]
@TomElliottFlexion @rgoshen @shumway-tm 

I started testing this story with the FF set to OFF. I found a security issue.  I notice that even with the FF set to OFF, private practitioners can still see Petitioner address/contact information in the Case JSON.  I compared this behavior in Staging, and in Staging, Petitioners cannot see Petitioner address/contact info in the JSON.

**Comment by ttlenard** [comment]
@TomElliottFlexion @rgoshen @shumway-tm 
Some more testing feedback:

- As a docket clerk, I removed a private practitioner from a consolidated case. It was the only case in the group that this practitioner had access to.
- Next, Log in as the same private practitioner, and navigate to a case in the group (can be the same case or a different case in the group)
- Click on the request access to case button
- Practitioner only sees a white screen.  
- If you pull up the console, there is a type error:

![image.png](https://images.zenhubusercontent.com/60f99b8ac0d8d04c8aed1aa3/143375b1-e2a9-4254-b9a7-5179b8dd1082)


This seems somewhat similar to the issue that I found while testing 9595.  See the comments.  A little bit different error, but similar workflow that produces an error (docket clerk removing a practitioner from a case in a group).
